shell/platform/android/android_image_generator.cc

```diff
@@ -0,0 +1,203 @@
+    unsigned int frame_index) const {
+  return {.required_frame = std::nullopt,
+          .duration = 0,
+          .disposal_method = SkCodecAnimation::DisposalMethod::kKeep};
+}
+
+SkISize AndroidImageGenerator::GetScaledDimensions(float desired_scale) {
+  return GetInfo().dimensions();
+}
+
+bool AndroidImageGenerator::GetPixels(const SkImageInfo& info,
+                                      void* pixels,
+                                      size_t row_bytes,
+                                      unsigned int frame_index,
+                                      std::optional<unsigned int> prior_frame) {
+  fully_decoded_latch_.Wait();
+
+  if (kRGBA_8888_SkColorType != info.colorType()) {
+    return false;
+  }
+
+  switch (info.alphaType()) {
+    case kOpaque_SkAlphaType:
+      if (kOpaque_SkAlphaType != GetInfo().alphaType()) {
+        return false;
+      }
+      break;
+    case kPremul_SkAlphaType:
+      break;
+    default:
+      return false;
+  }
+
+  // Since this decoder is only installed for API 28 and 29, there is no
+  // supported way to work around this unfortunate copy. It's possible to
+  // produce an `SkImage` direcly from an `AHardwareBuffer`, but
+  // `AndroidBitmap_getHardwareBuffer` and `Bitmap.getHardwareBuffer` are only
+  // available in API 30+ and 31+ respectively.
+  memcpy(pixels, software_decoded_data_->data(),
+         software_decoded_data_->size());
+  return true;
+}
+
+void AndroidImageGenerator::DecodeImage() {
+  DoDecodeImage();
+
+  header_decoded_latch_.Signal();
+  fully_decoded_latch_.Signal();
+}
+
+void AndroidImageGenerator::DoDecodeImage() {
+  FML_DCHECK(g_flutter_jni_class);
+  FML_DCHECK(g_decode_image_method);
+
+  // Call FlutterJNI.decodeImage
+
+  JNIEnv* env = fml::jni::AttachCurrentThread();
+
+  fml::jni::ScopedJavaLocalRef<jobject> direct_buffer(
+      env, env->NewDirectByteBuffer(const_cast<void*>(data_->data()),
+                                    data_->size()));
+
+  fml::jni::ScopedJavaGlobalRef<jobject>* bitmap =
+      new fml::jni::ScopedJavaGlobalRef(
+          env, env->CallStaticObjectMethod(g_flutter_jni_class->obj(),
+                                           g_decode_image_method,
+                                           direct_buffer.obj(), (long)this));
+  FML_CHECK(fml::jni::CheckException(env));
+
+  if (bitmap->is_null()) {
+    return;
+  }
+
+  AndroidBitmapInfo info;
+  int status;
+  if ((status = AndroidBitmap_getInfo(env, bitmap->obj(), &info)) < 0) {
+    FML_DLOG(ERROR) << "Failed to get bitmap info, status=" << status;
+    return;
+  }
+  FML_DCHECK(info.format == ANDROID_BITMAP_FORMAT_RGBA_8888);
+
+  // Lock the android buffer in a shared pointer
+
+  void* pixel_lock;
+  if ((status = AndroidBitmap_lockPixels(env, bitmap->obj(), &pixel_lock)) <
+      0) {
+    FML_DLOG(ERROR) << "Failed to lock pixels, error=" << status;
+    return;
+  }
+
+  SkData::ReleaseProc on_release = [](const void* ptr, void* context) -> void {
```

Comment:
Do we need to do anything to make sure that this runs on the correct thread?

Reply:
Good question -- I think so. In practice, this currently always happens on a different thread. No explicit mentioning of threading in the docs for `AndroidBitmap_unlockPixels`, though.

Reply:
SkData is thread safe.

Reply:
And if we end up on a different thread, AttachCurrentThread will make sure we're allowed to make JNI calls on it. Or reuse the JNIEnv for that thread if it already attached.